### ISSUE TITLE
POP OS 22.04 Kernel 5.17 fails to boot BPF errors and cryptsetup waiting for UUID

### ISSUE THREAD

**Issue by ganakee** [pop-os/pop#2387]
**Distribution (run `cat /etc/os-release`):**
cat /etc/os-release
NAME="Pop!_OS"
VERSION="22.04 LTS"
ID=pop
ID_LIKE="ubuntu debian"
PRETTY_NAME="Pop!_OS 22.04 LTS"
VERSION_ID="22.04"
HOME_URL="https://pop.system76.com"
SUPPORT_URL="https://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=jammy
UBUNTU_CODENAME=jammy
LOGO=distributor-logo-pop-os
 
**Related Application and/or Package Version (run `apt policy $PACKAGE NAME`):**

Kernel 5.17.5

**Issue/Bug Description:**
After restoring POP OS 22.04 from Timeshift backup, cannot boot into the 5.17.5 kernel. On boot, numerous shell errors scroll BPF type_id=78 and then the boot fails at cryptsetup waiting for UUID xxxxxxxx

Cannot boot into 5.17.5

**Steps to reproduce (if you know):**
Boot.

**Expected behavior:**
Boot into 5.17.5

**Other Notes:**
I have tried numerous fixes including restore from a current Timeshift backup, running update-initramfs -k all -c (with -u), checking the UUID of the swap partition (matches), reinstalling the kernel files via apt, checking /etc/crypttab etc. No matter hat I do, I can boot into the older 5.16 kernel fine but not the new 5.17.5 kernel. Tried several apt command sudo apt-get update --fix-missing, cleean autoremove, etc. Nothing helps.

Output from lsblk -f (removed part of UUIDs with ... but a full UUID appears in the output:
sdb                                                                         
├─sdb1
│    ext4   1.0         53b...d1bb8                
├─sdb2
│    ext4   1.0   /home 17c....44a2  334.7G    25% /home
├─sdb3
│    swap   1           285...d3e                
│ └─cryptswap
│    swap   1     cryptswap
│                       fe07...0071                [SWAP]
├─sdb4
│                                                                           
└─sdb5
     ext4   1.0         358...165eae3  134.7G    29% /

**Comment by ganakee** [comment]
Also tried reinstaling the linux-kernel-generic files and update-grub. Nothing helps.

**Comment by ganakee** [comment]
This is maddening. Why can't we easily refresh or re-install a broken kernel? I have tried every option that I can find and multiple restores from Timeshift (14 times). Nothing works. Every time I try to boot into the newest kernel 5.17.5, I get BPA offset errors, I get a BusyBox (initramfs) prompt (but cannot type), and then the system says it cannot load cryptfs at ???UUID and the stops and drops to a BusyBox screen (and freezes). I have waited up to 20 minutes for something to happen. I cannot uninstall the broken 5.17.5 packages (I get a fatal dependency error with linux-system76). I tried a forced apt install --reinstall the 5.17.5 related packages. Multiple update-initramfs. Nothing works to resolve, what I think should be, a simple issue--just refresh or regenerate the working version of the now-broken kernel. I can boot into an older kernel but would really like to figure out how to fix this issue. I have run fsck on the drives (no errors), tried gnome-disks check and repair drives. Nothing helps. I even went back and installed older 22.04 Timeshift images (with reinstall grub2, regenerate initramfs, and redo the grub menus. Do I need to do manual boots (this is a BIOS boot, not uefi) until a new kernel arrives and hope that fixes this issue?

**Comment by ganakee** [comment]
Thank you. 21.10

**Comment by ganakee** [comment]
I continue to work on this mystery.

I have tried modifying the grub menu to boot the drive /dev/sdb5 rather tha UUID etc.

One thing found today is that after running sudo update-initramfs -k 5.17.5-76051705-generic -u
the initraamfs seems to generate fine. When I look at /boot I see:

ls -la
total 509340
drwxr-xr-x  3 root root      4096 May 14 08:35 .
drwxr-xr-x 19 root root      4096 Apr 26 12:10 ..
-rw-r--r--  1 root root    261798 Mar 22 11:47 config-5.16.15-76051615-generic
-rw-r--r--  1 root root    261824 Apr 11 12:56 config-5.16.19-76051619-generic
-rw-r--r--  1 root root    263839 May  2 11:20 config-5.17.5-76051705-generic
drwxr-xr-x  6 root root      4096 May 14 08:35 grub
lrwxrwxrwx  1 root root        35 May 10 08:10 initrd.img -> initrd.img-5.16.19-76051619-generic
-rw-r--r--  1 root root 113527572 May  9 23:01 initrd.img-5.16.15-76051615-generic
-rw-r--r--  1 root root 113528564 May  9 23:01 initrd.img-5.16.19-76051619-generic
-rw-r--r--  1 root root 115028687 May 14 08:35 initrd.img-5.17.5-76051705-generic
lrwxrwxrwx  1 root root        34 May 14 08:35 initrd.img.old -> initrd.img-5.17.5-76051705-generic
-rw-r--r--  1 root root 115029772 May 10 08:08 initrd.img.sb.old
-rw-------  1 root root   6259037 Mar 22 11:47 System.map-5.16.15-76051615-generic
-rw-------  1 root root   6259874 Apr 11 12:56 System.map-5.16.19-76051619-generic
-rw-------  1 root root   6318509 May  2 11:20 System.map-5.17.5-76051705-generic
lrwxrwxrwx  1 root root        32 May 10 08:09 vmlinuz -> vmlinuz-5.16.19-76051619-generic
-rw-------  1 root root  11147936 Mar 22 11:47 vmlinuz-5.16.15-76051615-generic
-rw-------  1 root root  11154208 Apr 11 12:56 vmlinuz-5.16.19-76051619-generic
-rw-------  1 root root  11234720 May  2 11:20 vmlinuz-5.17.5-76051705-generic
lrwxrwxrwx  1 root root        31 May 14 08:35 vmlinuz.old -> vmlinuz-5.17.5-76051705-generic
-rw-------  1 root root  11234720 May 10 08:07 vmlinuz.sb.old

The initrd.img-5.17... seems to be new with the current date. But vmlinuz-5.17.5... seems to stay at 2022-05-02 date as do the corresponding files. I then ran sudo apt install --reinstall linux-image-5.17.5-76051705-generic to reinstall the image files. The dates stay the same.

The problem is nothing works to allow the 5.17.5 kernel to boot on my system. I get a series of errors BTF: -22 failed to validate module errors and ultimately, if I wait about 10 minutes on boot, several BusyBox messages with ALERT! UUIS 358... does not exist. Dropping to shell. 

But there is no input possible on the shell. Even if I direct plug a keyboard (not in a hub), no input.

I can boot into any older kernel.

**Comment by Seibz** [comment]
If you boot up the machine while holding down the spacebar, may we have you hit d on the currentkern option?

Does re-setting that as default get you back into a booting state?

**Comment by ganakee** [comment]
I have spent more than 15 hours on this because I really want to better understand why this is happening (and hoping a new kernel will somehow fix the issue).
 Nothing corrects the issue with 5.17.5. In addition to the above, I used the bootconfig script to see if there was an issue. I booted into the LIVE CD many times. The system appears to "see" the  BIOS MBR boot and then finds core.img. The failure seems to be with initramfs. I have re-installed the linux-general and linux-5.17x packages several times (from in LIVE CD and verifying the correct chrooted environment. Grub has been reinstalled several times and using grub-customizer to copy and try variants of the boot config. Every other 5.16 kernel loads fine. The 5.17.5 booted OK before the restore (several problems with sound and WIFI triggered the need to go back temporarily and then restore to 22.04).

This 5.17.5 kernel simply will not boot which really has caused me to rethink POPOS (not trying to be mean but more than 15 hours to try to fix without any results and no real explanation of why this is happening raises concerns about whther the problem will persist even if a new kernel gets released). I just get multiple offset errors when trying to boot 5.17.5, then cryptswap tries to load (so it gets to the POP OS grey screen), then the system drops to (INTRAMFS) and BUSYBOX (with no keyboard support) seeming to say the root drive cannot be found. (POP updates through 2022-05-17 have not helped.) The root drive UUID is correct, the location (physically on the disk) is OK, fsck shows no errors or issues, GNOME Disks repair shows no errors, all other POP OS kernels load fine, reinstalling kernel files, reinstalling grub, etc have no effect on the problem.

I realize that wiser people would just reinstall a clean version. My concern is whether a clean install will actually help. The 5.17 kernel caused the initial issues for me (lost WIFI, lost video, and sound issues) requiring the roll-back to 21.10 via TimeShift so I could work (I hate being dependent on video meetings but that is reality and losing video and audio quirkiness). 

Just puzzled about what may be happening and why apt install --reinstall, even when chrooted via LIVE CD and showing that a reinstall happened, does not reinstall the initramfs and other files. It just seems like me initramfs image file is somehow broken.

**Comment by ganakee** [comment]
I reviewed a fresh copy of bootinfoscript output this morning. (After running many repair attempts this weekend.) 

I finally ran boot-repair. That did not correct the issue.

But I do see in bootinfoscript output
 Boot sector info:  Grub2 (v1.99-2.00) is installed in the boot sector of 
                       sdb5 and looks at sector 1651161064 of the same hard 
                       drive for core.img, but core.img can not be found at 
                       this location.
I wanted to think that aha, this is the issue. I use gparted to see where the dedicated BIOS 1GB partition (GPT) resides. That is /sdb4 on my system but starts at 1515089920. Gparted shows that core.img resides there. But that appears to conflict, even after boot-repair, update-initramfs, etc. repairs pointing to 1651161064. However, I am not sure this is relevant at all because the system boots to all kernels BUT 5.17. I cannot see how to fix this issue (assusing it even is part of the issue, but why do all other kernels boot).

**Comment by ganakee** [comment]
The POP OS updates from 2022-05-27 including a revised release of vmlinuz image, revised kernel modules, and a re-generated kernel fixed this persistent issue. I can now, after installing those updates, boot into 5.17.5 again.

uname -r
5.17.5-76051705-generic

**Comment by RinusHdev** [comment]
Just installed POP OS in place of an Ubuntu (21.10, I think) installation and I think I encountered a similar issue.
During installation:
1. Indicated the old /boot ext4 partition as /boot and to be formatted
2. Indicated the old / ext4 partition as / and to be formatted
3. Using Gparted from the installer shrunk the old encrypted LUKS (containing /home) partition to make room for a swap partition
4. Added a swap partition in Gparted
5. Marked the swap partition as swap
6. Marked the ext4 partition in the encrypted partition as the new /home, not to be formatted.

While performing step 56 the installer errored decrypting the LUKS partition as "cryptdata" was already taken. Probably due to closing the encryption from within gparted instead of the installer? Changing that to "cryptdata2" allowed the process to continue. Installation finished, reboot... emergency mode.

/etc/fstab contained the right UUID for the /home "partition" (forgive me the term if its actually called a map or something) within the encrypted partition.
/etc/crypttab, however, only contained the entry for the swap partition, not for the luks-encrypted partition. blkid, nor lsblk showed the encrypted home partition within, only the "container". 
Adding an entry to /etc/crypttab fixed the issue for me.

What in the end caused the missing entry, I don't know, but there seems to be a possibility for error on ugrading from Ubuntu where encrypted volumes are concerned.
